spmd/compiler/api.py

```diff
@@ -171,11 +173,34 @@ def dummy_add(grad: torch.Tensor, zero: torch.Tensor) -> torch.Tensor:
         call_functions[0], node_to_obj
     )
 
+    # rewrite graph to remove redundant clone
+    traced_dispatch = _remove_clone_tensor(traced_dispatch)
+
     traced_dispatch.graph.lint()
 
     return traced_dispatch
 
 
+def _remove_clone_tensor(subgm: fx.GraphModule) -> fx.GraphModule:
```

Comment:
A few things we can do here to make it more generic:

0. make sure the original node (clone's arg) are not called by any inplace ops after clone and before used in allreduce. Otherwise, linking allreduce to the original node might be wrong. 
1. relax the comm node from allreduce to other types of node as well. The tricky thing here is how to deal with inplace ops. If there are ops using the original node after allreduce (or other inplace ops), removing clone will make the behavior different.  
2. call `eliminate_dead_code()` to automatically remove the clone node.

Reply:
Not sure if the following is correct:

For allreduce (or other ops) to replace its clone arg with the original arg: there can be any inplace op before allreduce and after clone

For subsequent logic to be correct: there can be any op (inplace or not inplace) after allreduce using the original node. Otherwise, allreduce would have modified value.

Reply:
couple of questions:
1. For the 0th point above, given that we are adding the dummy_add operation is there any danger of other ops in between the clone and allreaduce? If yes, what is an example of that.
2. A noob question: what does this optimization give us in terms of performance or usability?

Reply:
For item 0:
I've added a user check to confirm there is only one use of the grad_tensor in the comm section subgraph. 
<img width="1019" alt="grad_tensor_one_user" src="https://user-images.githubusercontent.com/46302957/204431973-916187e1-946b-4dc1-afa5-75e68f3beb1a.png">

Reply:
Related - for @anj-s  item 1 - the check/assert added above confirms that there will only be 1 user in the subgraph, so that should remove this issue of concern re: other users of the clone.

For item 2 - what perf or usability gains?
Main gain is in perf - the clone nodes are duplicating every grad tensor in the backward graph on cpu.  There is a small hit for each cpu alloc perf wise, and esp for for larger models (think 170B) we could end up blowing out or at least consuming a large amount of cpu memory trying to clone every tensor in the graph as it currently does (not sure how fast gc would remove these vs. creation).  
I don't think there is any usability gain other than we have a cleaner graph to work with.  The main gain imo is that we are not spinning up a large part of an entire model clone on the users cpu and assuming that is a blocking mem alloc, should improve perf a bit relative to the size of the clones being created.

Reply:
related - additionally for perf, the gradients would need to flow from the clone back to the source tensor, so that is additional overhead involved as well as the aforementioned memory alloc time and overhead.  

Final item - I'm unclear why using a cpu based clone does not cause an error since this is using NCCL. But it is another reason we want to remove these imo and use the gpu gradient tensors directly.